include/swift/AST/DiagnosticEngine.h

```diff
@@ -672,6 +672,19 @@ namespace swift {
       return diagnose(Loc, Diagnostic(ID, std::move(Args)...));
     }
 
+    /// \brief Emit a diagnostic with the given set of diagnostic arguments.
+    ///
+    /// \param ID The diagnostic to be emitted.
+    ///
+    /// \param Args The diagnostic arguments, which will be converted to
+    /// the types expected by the diagnostic \p ID.
+    template<typename ...ArgTypes>
+    InFlightDiagnostic
+    diagnose(Diag<ArgTypes...> ID,
```

Comment:
We've deliberately not added this overload because it'd be too easy to invoke by accident. Maybe we can use some template hackery to limit it to certain diagnostics?

Reply:
hmm, wouldn't this misuse be caught by any diagnostics tests that are source location specific?

Reply:
They ought to be, sure, but when not testing with `-verify` we could easily miss it. It's also nice if the compiler can just tell you.

Reply:
ok, i'll remove this overload and add `SourceLoc()` at the callsites.